cpp/ql/src/Likely Bugs/Format/NonConstantFormat.ql

```diff
@@ -68,9 +68,9 @@ predicate cannotContainString(Type t, boolean isIndirect) {
 
 predicate isNonConst(DataFlow::Node node, boolean isIndirect) {
   exists(Expr e |
-    e = node.asExpr() and isIndirect = false
+    e = [node.asExpr(), node.asDefinition()] and isIndirect = false
     or
-    e = node.asIndirectExpr() and isIndirect = true
+    e = [node.asIndirectExpr(), node.asIndirectDefinition()] and isIndirect = true
```

Comment:
You're example I cannot quite follow. I would have expected that the `hello` argument of `printf` uses the value after the increment, and that suffices, because we're looking at the argument of `printf` when looking for sinks.

Reply:
Sorry. What I meant was there are no instruction that holds the result of the operation. There is an `AddInstruction` and a `StoreInstruction`. But there's no `CopyValueInstruction` that holds the old value. i.e., if we have:
```cpp
int x = 0;
int y = x++;
```
then there will be a `CopyValueInstruction` whose `getUnconvertedResultExpression()` is the `CrementOperation` so that it can be used in the `StoreInstruction` that writes to `y`:
```cpp
r5_1(glval<int>) = VariableAddress[x] :
r5_2(int)        = Constant[0]        :
m5_3(int)        = Store[x]           : &:r5_1, r5_2
r6_1(glval<int>) = VariableAddress[y] :
r6_2(glval<int>) = VariableAddress[x] :
r6_3(int)        = Load[x]            : &:r6_2, m5_3
r6_4(int)        = Constant[1]        :
r6_5(int)        = Add                : r6_3, r6_4
m6_6(int)        = Store[x]           : &:r6_2, r6_5
r6_7(int)        = CopyValue          : r6_3
m6_8(int)        = Store[y]           : &:r6_1, r6_7
```
But if the increment occurs in a void context such as:
```cpp
int x = 0;
x++;
```
then the IR is:
```cpp
r5_1(glval<int>) = VariableAddress[x] :
r5_2(int)        = Constant[0]        :
m5_3(int)        = Store[x]           : &:r5_1, r5_2
r6_1(glval<int>) = VariableAddress[x] :
r6_2(int)        = Load[x]            : &:r6_1, m5_3
r6_3(int)        = Constant[1]        :
r6_4(int)        = Add                : r6_2, r6_3
m6_5(int)        = Store[x]           : &:r6_1, r6_4
```
and so there's no `CopyValueInstruction` generated. So there's no instruction for which `getUnconvertedResultExpression` gives the crement operation. And that's why `node.asExpr()`.

So what I'm suggesting now is that, if we're in the latter case where there is no instruction that gives back the crement operation, then `node.asExpr()` should behave as `node.asDefinition()` and give back the post incremented value.

Reply:
> My understanding is that CopyValueInstructions are just a happy accident of how we generate the IR, and behaviour should not change when we omit them. Is that wrong?

They're not always happy accidents. For some cases (including the case we're discussing here: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll#L3264-L3269) we synthesize a `CopyValueInstruction` precisely so that there's an instruction to hold the result of the expression. But we don't do so when it's in a void context.

Reply:
> So what I'm suggesting now is that, if we're in the latter case where there is no instruction that gives back the crement operation, then `node.asExpr()` should behave as `node.asDefinition()` and give back the post incremented value.

I've done this in 03b77dbf2a46e63e5e5171a0323af95b259e9836. This actually means that we don't need to special-case `asDefinition` in all the place where we had to special-case it before. Now, this will Just Work:
```cpp
const char *hello = "Hello, World\n";
hello++;
printf(hello); // BAD
```

Reply:
I expect the result of the load to be the thing that is tainted here and picked up, but that's apparently not what's going on given that we're somehow talking about the increment operation here.

Reply:
The source in the non-constant format query is the increment operation. That's why we're talking about that operation here, no? 😂 And the load is loading from the address produced by the increment operation.

Reply:
Ahhh.

I see why I got confused. We're fixing something that affects duplication at sinks, but somehow - which I had not understood - it also affects sources.

Reply:
Ok, so doing the special handling in void contexts - as you do now - is probably the least confusing way to deal with this from a user perspective. Avoids having to learn about definitions being special most of the time.

Reply:
Yeah, I think I'm actually pretty happy with this now. Now you'll only need to deal with `asDefinition()` when there's actually a difference between what you'd get from calling `asExpr()` and `asDefinition()`.